internal/cephfs/clone.go

```diff
@@ -0,0 +1,226 @@
+		return cloneFailedErr
+	case cephFSCloneComplete:
+		// This is a work around to fix sizing issue for cloned images
+		err = resizeVolume(ctx, volOpt, cr, cloneID, volOpt.Size)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to expand volume %s: %v"), cloneID, err)
+			return err
+		}
+		// As we completed clone, remove the intermediate snap
+		if err = unprotectSnapshot(ctx, parentvolOpt, cr, snapshotID, volID); err != nil {
+			// Incase the snap is already unprotected we get ErrSnapProtectionExist error code
+			// in that case we are safe and we could discard this error and we are good to go
+			// ahead with deletion
+			if !errors.Is(err, ErrSnapProtectionExist) {
+				klog.Errorf(util.Log(ctx, "failed to unprotect snapshot %s %v"), snapshotID, err)
+				return err
+			}
+		}
+		if err = deleteSnapshot(ctx, parentvolOpt, cr, snapshotID, volID); err != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapshotID, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanupCloneFromSubvolumeSnapshot(ctx context.Context, volID, cloneID volumeID, parentVolOpt *volumeOptions, cr *util.Credentials) error {
+	// snapshot name is same as clone name as we need a name which can be
+	// identified during PVC-PVC cloning.
+	snapShotID := cloneID
+	snapInfo, err := getSnapshotInfo(ctx, parentVolOpt, cr, snapShotID, volID)
+	if err != nil {
+		if errors.Is(err, ErrSnapNotFound) {
+			return nil
+		}
+		return err
+	}
+
+	if snapInfo.Protected == snapshotIsProtected {
+		err = unprotectSnapshot(ctx, parentVolOpt, cr, snapShotID, volID)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to unprotect snapshot %s %v"), snapShotID, err)
+			return err
+		}
+	}
+	err = deleteSnapshot(ctx, parentVolOpt, cr, snapShotID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapShotID, err)
+		return err
+	}
+	return nil
+}
+
+func createCloneFromSnapshot(ctx context.Context, parentVolOpt, volOptions *volumeOptions, vID *volumeIdentifier, sID *snapshotIdentifier, cr *util.Credentials) error {
+	snapID := volumeID(sID.FsSnapshotName)
+	err := cloneSnapshot(ctx, parentVolOpt, cr, volumeID(sID.FsSubvolName), snapID, volumeID(vID.FsSubvolName), volOptions)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			if !errors.Is(err, ErrCloneInProgress) {
+				if dErr := purgeVolume(ctx, volumeID(vID.FsSubvolName), cr, volOptions, true); dErr != nil {
+					klog.Errorf(util.Log(ctx, "failed to delete volume %s: %v"), vID.FsSubvolName, dErr)
+				}
+			}
+		}
+	}()
+	var clone = CloneStatus{}
+	clone, err = getCloneInfo(ctx, volOptions, cr, volumeID(vID.FsSubvolName))
+	if err != nil {
+		return err
+	}
+	switch clone.Status.State {
+	case cephFSCloneInprogress:
+		return ErrCloneInProgress
+	case cephFSCloneFailed:
+		return fmt.Errorf("clone %s is in %s state", vID.FsSubvolName, clone.Status.State)
+
```

Comment:
why an empty line, if you're not doing that for the other switch/case?

Reply:
done.